testing/projects/cpp_nesting/include/nesting_specializations.hpp

```diff
@@ -0,0 +1,176 @@
+    template <class TImage, typename Tdata_t, int Tchannels>
+    class ImageBuffer {
+    public:
+        using Image_t = TImage;
+        using data_t = Tdata_t;
+        static constexpr int width = Image_t::width;
+        static constexpr int height = Image_t::height;
+        static constexpr int channels = Tchannels;
+
+        /// Never do this in real life...
+        struct Data {
+            data_t vals[channels];///< The pixel data.
+        };
+
+        /// Checking nested specials too.
+        template <int TNested>
+        struct SomeThing {
+            std::string id() const { return std::to_string(TNested); }
+        };
+
+        /// How much memory is used.
+        constexpr int size() const {
+            return width * height * sizeof(Data);
+        }
+    };
+
+    /// Now lets specialize it partially and arbitrarily change behavior!
+    template <class TImage, typename Tdata_t>
+    class ImageBuffer<TImage, Tdata_t, 4> {
+    public:
+        using Image_t = TImage;
+        using data_t = Tdata_t;
+        static constexpr int width = Image_t::width;
+        static constexpr int height = Image_t::height;
+        static constexpr int channels = 4;
+
+        /// So really, don't change how things work like this in real life...
+        struct Data {
+            data_t r;///< Red.
+            data_t g;///< Green.
+            data_t b;///< Blue.
+            data_t a;///< Alpha.
+        };
+
+        /// This one is not as special, right?
+        struct SomeThing {
+            std::string id() const { return "I AM SPECIAL!"; }
+        };
+
+        /// How much memory is used.
+        constexpr int size() const {
+            return width * height * 4 * sizeof(data_t);
+        }
+    };
+
+    /// A fully specialized buffah.
+    template <>
+    class ImageBuffer<Image<1920, 1080>, float, 128> {
+    public:
+        using Image_t = Image<1920, 1080>;
+        using data_t = float;
+        static constexpr int width = 1920;
+        static constexpr int height = 1080;
+        static constexpr int channels = 128;
+
+        /// Definitely don't do this.
+        struct Data {
+            std::vector<float> samples;///< so arbitrary
+        };
+
+        /// Super special.
+        struct SomeThing {
+            std::string id() const { return "Spectral"; }
+        };
+
+        /// This is a lie.
+        constexpr int size() const { return 88; }
+    };
+
+#if 0
+    /** The original example expanded a bit. Thanks florian for making it
+     * copy-pasteable :) */
+    template <size_t N>
+    struct Base {
+        /// How big the base is.
+        constexpr size_t size() const { return N; }
+        struct A {};///< eeep
+    };
+
+    /// Specialized base.
+    template <>
+    struct Base<2> {
+        /// It is bigger for no real reason.
+        constexpr size_t size() const { return 42; }
+
+        /// Inner struct 1.
+        struct InnerStruct {
+            int a() const { return 3; };
```

Comment:
the only reason they are functions really is so that the code coverage tool actually hits a line xD  it really doesn't matter much